### DISCUSSION TITLE
Use auth.uid() to prevent forged user_ids?

### DISCUSSION THREAD

**Discussion by city41** [supabase/supabase#802] [Questions]
In the slack clone example, whenever someone says something, this insert call is made

```javascript
export const addMessage = async (message, channel_id, user_id) => {
  try {
    let { body } = await supabase.from('messages').insert([{ message, channel_id, user_id }])
    return body
  } catch (error) {
    console.log('error', error)
  }
}
```

[link to the code](https://github.com/supabase/supabase/blob/master/examples/nextjs-slack-clone/lib/Store.js#L159)

But since the client is setting `user_id`, it's possible to swap in a different id and make other users "say" something. I confirmed this by hardcoding user_id to supabot's id and from then on all messages in the channel came "from" supabot.

I believe I was able to prevent this attack by adding this trigger

```sql
CREATE OR REPLACE FUNCTION "server_side_user_id"()
  RETURNS trigger
  AS $pga$
BEGIN
  NEW.user_id = auth.uid();
  return NEW;
END;
    $pga$
  VOLATILE
  LANGUAGE plpgsql security definer;

CREATE or replace TRIGGER "hash_level_data_trigger"
  before insert ON "messages"
  FOR EACH row
  EXECUTE PROCEDURE "server_side_user_id"();
```

And with this trigger in place, the client can no longer hack user_id. 

Is this a good way to go about this? Are there better ways to solve this?

**Comment by steve-chavez** [comment]
Hm, that is weird because the slack clone schema has this RLS policy:

```sql
create policy "Allow individual insert access" on public.messages
for insert with check ( auth.uid() = user_id );
```

(reference: https://github.com/supabase/supabase/tree/master/examples/nextjs-slack-clone#postgres-row-level-security)

Which should only allow a user to insert messages for himself and not for other users.

Can you double check RLS is enabled for the messages table?

**Reply by city41** [reply]
I noticed that as well, but at least in my testing it doesn't seem to be working. I'm not very familiar with RLS though. If that policy is the ideal way to accomplish this that is good to know, I can explore why it either isn't working or perhaps I just made a bad assumption somewhere.

and yeah, the policy does look to be in place

![image](https://user-images.githubusercontent.com/141159/109995103-f846be80-7cdb-11eb-990e-dc7b02d9b130.png)

**Reply by city41** [reply]
From what I can tell, no policies work on this table at all. Even really draconian ones, here are some I tried, each commented out line was tried, with and without `for insert`

```sql
ALTER TABLE public.messages enable ROW LEVEL SECURITY;
ALTER TABLE public.messages FORCE ROW LEVEL SECURITY;

drop policy if exists no_insert_forge on public.messages;

create POLICY no_insert_forge ON public.messages
    --for insert
    --with check (user_id = auth.uid())
    --with check (user_id = '8d0fd2b3-9ca7-4d9e-a95f-9e13ddedaaaa');
    --with check (message = 'must be this');
    with check (false);
```

Reading the docs, force row level security is needed to enforce it on the table owner. The owner is supabase_admin, so that doesn't seem relevant.

But when I do this

```sql
select schemaname, tablename, rowsecurity from pg_tables where tablename = 'messages';
```

I get back

![image](https://user-images.githubusercontent.com/141159/109999184-f252dc80-7cdf-11eb-9112-2eb90ba0548b.png)


Which suggest RLS Is not actually enabled?

**Reply by steve-chavez** [reply]
> Which suggest RLS Is not actually enabled?

Can you double click on `rowsecurity`? I think the boolean value is being greyed now(bug).

Also, could you check with what role you're making the requests?(what's the value of the role claim in the JWT)

This might help with that:
```sql
create or replace function show_current_user() returns name as $$
  select current_user;
$$ language sql; 
```

And then call the function it through `supabase.rpc("show_current_user")`.

**Reply by city41** [reply]
ah you're right, double clicking the cell shows "true".

I added that function, then added a call to it on the client

```javascript
export const addMessage = async (message, channel_id, user_id) => {
  // user_id = '4eafa20e-3365-4895-b525-a44395d4202f';
  try {
    let result = await supabase.rpc("show_current_user");
    console.log('result', result);
    let { body } = await supabase.from('messages').insert([{ message, channel_id, user_id }])
    return body
  } catch (error) {
    console.log('error', error)
  }
}
```

and the result is "authenticated". I see there is an authenticated role in the db, but from what I can tell it's not capable of much.

**Reply by steve-chavez** [reply]
Hm, authenticated role looks fine. I thought it might be a superuser and that one would bypass RLS.

Just in case you could double check that both `rolsuper` and `rolbypassrls` below are false:
```sql
select rolname, rolsuper, rolbypassrls from pg_roles where rolname = 'authenticated';
```

**Reply by city41** [reply]
Yup, both are false for that query.

Something def seems off. I haven't done anything out of the ordinary, just stood up the slack clone. This is starting to feel like worthy of being posted over in issues.

**Reply by steve-chavez** [reply]
> This is starting to feel like worthy of being posted over in issues.

Yes, definitely. We'd really appreciate an issue to track this better.

Btw, the trigger workaround is good. Though ideally the RLS policy should disallow forging the user id so there wouldn't be a problem in setting it client-side.

**Reply by city41** [reply]
ok, issue created here https://github.com/supabase/supabase/issues/810

thanks for all your help!
